### ISSUE TITLE
KNN evaluation failing for CIFAR10 on ImageNet pre-trained model.

### ISSUE THREAD

**Issue by reyanshsolis** [facebookresearch/dino#52]
Dear Authors,

Thank you for this amazing work and repository. I wanted to see the clustering capability of the pretrained network on imagenet on CIFAR-10 dataset. For doing so I wanted to use the features you extract in the eval_knn.py script. Below are the parameters I used to run the code:

```bash
arch= deit_small (i.e vit_small after the new commit)
patch_size=16
batch_size_per_gpu = 32
data_path = 'path_to_cifar10_dataset'
and all other params are set to their default values.
```

I left the `pretrained_weights` as blank so that the model loads the weights of imagenet from the url mentioned in `utils.py`

The script executed successfully with the message:
```bash
Data loaded with 50000 train and 10000 val imgs.
Model deit_small 16x16 built.
Please use the `--pretrained_weights` argument to indicate the path of the checkpoint to evaluate.
Since no pretrained weights have been provided, we load the reference pretrained DINO weights.
Pretrained weights found at dino_deitsmall16_pretrain/dino_deitsmall16_pretrain.pth and loaded with msg: <All keys matched successfully>
Extracting features for train set...
Storing features into tensor of shape torch.Size([50000, 384])
```
However, I'm facing a strange issue, the features extracted are coming out to be a zero vector, for all of the input image.
```bash
torch.Size([50000, 384])
torch.Size([50000])
tensor([[0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.],
        ...,
        [0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.]], device='cuda:0')
```

Can you please specify where I might be going wrong.
Thanks in advance!

**Comment by mathildecaron31** [comment]
Hi @reyanshsolis 

Thanks for your interest and your kind words.

The matrix used for storing the features is initialized to zero everywhere:
https://github.com/facebookresearch/dino/blob/4b96393c4c877d127cff9f077468e4a1cc2b5e2d/eval_knn.py#L95

so if you have printed it before the feature extraction it makes sense that it is zero everywhere.

Otherwise I don't know what happens... We do have experiments when transferring to CIFAR-10 and it was working fine.

**Comment by ZuowenWang0000** [comment]
Hi, @reyanshsolis I encountered a similar problem. The mistake I made is the evaluation script is using the ReturnIndexDataset class. So the expected return of the dataloader object is image, index instead of image, label. (if you do so, only 10 images' features will be loaded for cifar10). So I had to wrap the dataloader with the ReturnIndexDataset and it worked.
You also have to load all the labels separately, but that should be easy for cifar i guess.
